pkg/agent/stats/collector.go

```diff
@@ -0,0 +1,207 @@
+	antreaClusterNetworkPolicyStats map[types.UID]*statsv1alpha1.TrafficStats
+	// antreaNetworkPolicyStats is a mapping from Antrea NetworkPolicy UIDs to their traffic stats.
+	antreaNetworkPolicyStats map[types.UID]*statsv1alpha1.TrafficStats
+}
+
+// Collector is responsible for collecting stats from the Openflow client, calculating the delta compared with the last
+// reported stats, and reporting it to the antrea-controller summary API.
+type Collector struct {
+	nodeName string
+	// antreaClientProvider provides interfaces to get antreaClient, which will be used to report the statistics to the
+	// antrea-controller.
+	antreaClientProvider agent.AntreaClientProvider
+	// ofClient is the Openflow interface that can fetch the statistic of the Openflow entries.
+	ofClient openflow.Client
+	// lastStatsMap is the last statistics that has been reported to antrea-controller successfully.
+	// It is used to calculate the delta of the statistics that will be reported.
+	lastStatsCollection *statsCollection
+}
+
+func NewCollector(antreaClientProvider agent.AntreaClientProvider, ofClient openflow.Client) *Collector {
+	nodeName, _ := env.GetNodeName()
+	manager := &Collector{
+		nodeName:             nodeName,
+		antreaClientProvider: antreaClientProvider,
+		ofClient:             ofClient,
+	}
+	return manager
+}
+
+// Run runs a loop that collects statistics and reports them until the provided channel is closed.
+func (m *Collector) Run(stopCh <-chan struct{}) {
+	klog.Info("Start collecting metrics")
+	ticker := time.NewTicker(collectPeriod)
+	defer ticker.Stop()
+
+	// Record the initial statistics as the base that will be used to calculate the delta.
+	// If the counters increase during antrea-agent's downtime, the delta will not be reported to the antrea-controller,
+	// it's however better than reporting the full statistics twice which could introduce greater deviations.
+	m.lastStatsCollection = m.collect()
+
+	for {
+		select {
+		case <-ticker.C:
+			curStatsCollection := m.collect()
+			// Do not update m.lastStatsMap if the report fails so that the next report attempt can add up the
+			// statistics produced in this duration.
+			if err := m.report(curStatsCollection); err != nil {
+				klog.Errorf("Failed to report stats: %v", err)
+			} else {
+				m.lastStatsCollection = curStatsCollection
+			}
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+// collect collects the stats of Openflow rules, maps them to the stats of NetworkPolicies.
+// It returns a map from NetworkPolicyReferences to their stats.
+func (m *Collector) collect() *statsCollection {
+	// TODO: The following process is not atomic, there's a chance that the ofID is released and reused by another
+	//  NetworkPolicy rule in-between, leading to incorrect metrics. We should return relevant NetworkPolicy references
+	//  along with metrics to avoid it.
+	ruleStatsMap := m.ofClient.NetworkPolicyMetrics()
```

Comment:
should this be renamed from metrics to stats for consistency?

Reply:
I think we should, but since it doesn't affect user facing API and the method is not added by this PR, I plan to change it along with addressing the TODOs and ensuring its efficiency. Does it make sense?

Reply:
Sounds good